pandas/core/indexes/base.py

```diff
@@ -2256,15 +2256,15 @@ def intersection(self, other):
             indexer = indexer.take((indexer != -1).nonzero()[0])
         except:
             # duplicates
-            indexer = Index(other._values).get_indexer_non_unique(
-                self._values)[0].unique()
+            indexer = algos.unique1d(Index(other._values).get_indexer_non_unique(
```

Comment:
Reason I avoided that was because the behaviour of np unique is slightly different to the original unique function, it returns as sorted rather than the original order for example).